File: src/tools.py
"""Containerized tools with Docker sandboxing and fallback safety."""
import asyncio
import logging
from dataclasses import dataclass
from pathlib import Path
from typing import Optional, Tuple

from .sandbox import DockerSandbox, SandboxConfig, FallbackSafetyMode
from .safety import SafetyMiddleware
from .memory import RAGMemory

logger = logging.getLogger("linguclaw.tools")


@dataclass(frozen=True)
class CommandResult:
    stdout: str
    stderr: str
    returncode: int
    sandboxed: bool = True


@dataclass(frozen=True)
class FileResult:
    success: bool
    content: Optional[str] = None
    error: Optional[str] = None


class ShellTool:
    """Docker-sandboxed shell execution with fallback to strict safety mode."""

    def __init__(self, project_root: str, use_docker: bool = True, 
                 safety: Optional[SafetyMiddleware] = None,
                 fallback_confirmed: bool = False):
        self.project_root = Path(project_root).resolve()
        self.project_root.mkdir(parents=True, exist_ok=True)
        self.safety = safety or SafetyMiddleware()
        
        # Docker sandbox setup
        self.sandbox: Optional[DockerSandbox] = None
        self.fallback: Optional[FallbackSafetyMode] = None
        self._using_docker = False
        
        if use_docker:
            self.sandbox = DockerSandbox(SandboxConfig(
                image="alpine:latest",
                memory_limit="512m",
                cpu_limit=0.5,
                auto_remove=True,
            ))
            if self.sandbox.available:
                if self.sandbox.start(str(self.project_root)):
                    self._using_docker = True
                    logger.info("Using Docker sandbox")
                else:
                    logger.warning("Docker available but sandbox start failed")
                    self._init_fallback(fallback_confirmed)
            else:
                self._init_fallback(fallback_confirmed)
        else:
            self._init_fallback(fallback_confirmed)

    def _init_fallback(self, confirmed: bool):
        """Initialize strict safety fallback mode."""
        self.fallback = FallbackSafetyMode(self.safety, confirmed)
        logger.warning("Docker unavailable - using strict safety fallback")

    @property
    def is_sandboxed(self) -> bool:
        return self._using_docker and self.sandbox is not None

    async def run(self, command: str, timeout: float = 60.0) -> CommandResult:
        """Execute command in Docker container or with strict safety fallback."""
        
        if self.is_sandboxed and self.sandbox:
            # Docker sandboxed execution
            try:
                # Run in executor to make it async
                loop = asyncio.get_event_loop()
                exit_code, stdout, stderr = await asyncio.wait_for(
                    loop.run_in_executor(None, self.sandbox.exec, command),
                    timeout=timeout
                )
                return CommandResult(stdout, stderr, exit_code, sandboxed=True)
            except asyncio.TimeoutError:
                return CommandResult("", f"Timeout after {timeout}s", -1, sandboxed=True)
            except Exception as e:
                return CommandResult("", f"Sandbox error: {e}", -1, sandboxed=True)
        
        elif self.fallback:
            # Strict safety fallback - requires explicit confirmation
            allowed, reason = self.fallback.check(command)
            
            if not allowed:
                # Try to get explicit confirmation for this command
                if self.fallback.prompt_confirmation(command):
                    allowed = True
            
            if not allowed:
                return CommandResult("", f"STRICT SAFETY BLOCKED: {reason}", -1, sandboxed=False)
            
            # Execute with basic subprocess (only after confirmation)
            try:
                import subprocess
                proc = await asyncio.create_subprocess_shell(
                    command,
                    stdout=asyncio.subprocess.PIPE,
                    stderr=asyncio.subprocess.PIPE,
                    cwd=str(self.project_root),
                )
                stdout, stderr = await asyncio.wait_for(proc.communicate(), timeout=timeout)
                return CommandResult(
                    stdout.decode("utf-8", errors="replace"),
                    stderr.decode("utf-8", errors="replace"),
                    proc.returncode or 0,
                    sandboxed=False
                )
            except asyncio.TimeoutError:
                proc.kill()
                await proc.wait()
                return CommandResult("", f"Timeout after {timeout}s", -1, sandboxed=False)
            except Exception as e:
                return CommandResult("", str(e), -1, sandboxed=False)
        
        return CommandResult("", "No execution method available", -1, sandboxed=False)

    def stop(self):
        """Cleanup sandbox resources."""
        if self.sandbox:
            self.sandbox.stop()

class SearchMemoryTool:
    """RAG-based semantic search over codebase memory."""

    def __init__(self, memory: RAGMemory):
        self.memory = memory

    def search_codebase(self, query: str, k: int = 5) -> str:
        """Search the indexed codebase for relevant code."""
        if not self.memory.available:
            return "[Memory unavailable - RAG system offline]"
        return self.memory.search(query, k)

    def get_stats(self) -> dict:
        return self.memory.get_stats()

class FileSystemTool:
    """Sandboxed filesystem operations."""

    def __init__(self, root: str):
        self.root = Path(root).resolve()
        self.root.mkdir(parents=True, exist_ok=True)

    def _validate(self, path: str) -> Path:
        target = Path(path).resolve()
        if not target.is_relative_to(self.root):
            raise PermissionError(f"Access denied: {path} outside {self.root}")
        return target

    def read(self, path: str, limit: Optional[int] = None) -> FileResult:
        try:
            t = self._validate(path)
            if not t.exists(): return FileResult(False, error=f"Not found: {path}")
            if not t.is_file(): return FileResult(False, error=f"Not a file: {path}")
            c = t.read_text(encoding="utf-8", errors="replace")
            if limit and len(c) > limit: c = c[:limit] + f"\n... ({len(c)-limit} more)"
            return FileResult(True, content=c)
        except PermissionError as e: return FileResult(False, error=str(e))
        except Exception as e: return FileResult(False, error=f"Read error: {e}")

    def write(self, path: str, content: str) -> FileResult:
        try:
            t = self._validate(path)
            t.parent.mkdir(parents=True, exist_ok=True)
            t.write_text(content, encoding="utf-8")
            return FileResult(True)
        except PermissionError as e: return FileResult(False, error=str(e))
        except Exception as e: return FileResult(False, error=f"Write error: {e}")

    def list_dir(self, path: str = ".") -> FileResult:
        try:
            t = self._validate(path)
            if not t.exists(): return FileResult(False, error=f"Not found: {path}")
            if not t.is_dir(): return FileResult(False, error=f"Not a directory: {path}")
            entries = [f"{'[dir]' if e.is_dir() else '[file]'} {e.name}" for e in t.iterdir()]
            return FileResult(True, content="\n".join(sorted(entries)))
        except PermissionError as e: return FileResult(False, error=str(e))
        except Exception as e: return FileResult(False, error=f"List error: {e}")

File: src/tools.ts
/**
 * Containerized tools with Docker sandboxing
 * TypeScript equivalent of Python tools.py
 */

import path from 'path';
import fs from 'fs';
import { promisify } from 'util';
import { exec as execCallback } from 'child_process';
import { DockerSandbox } from './sandbox';
import { SafetyMiddleware, FallbackSafetyMode } from './safety';
import { CommandResult, FileResult, SandboxConfig } from './types';
import { getLogger } from './logger';

const execAsync = promisify(execCallback);
const logger = getLogger();

export class ShellTool {
  projectRoot: string;
  safety: SafetyMiddleware;
  sandbox: DockerSandbox | null;
  fallback: FallbackSafetyMode | null;
  private usingDocker: boolean;

  constructor(
    projectRoot: string,
    useDocker: boolean = true,
    safety?: SafetyMiddleware,
    fallbackConfirmed: boolean = false
  ) {
    this.projectRoot = path.resolve(projectRoot);
    this.safety = safety || new SafetyMiddleware();
    this.sandbox = null;
    this.fallback = null;
    this.usingDocker = false;

    // Ensure project directory exists
    if (!fs.existsSync(this.projectRoot)) {
      fs.mkdirSync(this.projectRoot, { recursive: true });
    }

    if (useDocker) {
      this.sandbox = new DockerSandbox({
        image: 'alpine:latest',
        memory_limit: '512m',
        cpu_limit: 0.5,
        auto_remove: true,
      });
    } else {
      this.initFallback(fallbackConfirmed);
    }
  }

  /**
   * Initialize Docker sandbox
   */
  async init(): Promise<boolean> {
    if (this.sandbox) {
      const available = await this.sandbox.checkAvailability();
      if (available) {
        const started = await this.sandbox.start(this.projectRoot);
        if (started) {
          this.usingDocker = true;
          logger.info('Using Docker sandbox');
          return true;
        }
      }
      logger.warn('Docker available but sandbox start failed');
    }
    
    this.initFallback(false);
    return false;
  }

  private initFallback(confirmed: boolean): void {
    this.fallback = new FallbackSafetyMode(this.safety, confirmed);
    logger.warn('Docker unavailable - using strict safety fallback');
  }

  /**
   * Check if running in sandbox
   */
  get isSandboxed(): boolean {
    return this.usingDocker && this.sandbox !== null;
  }

  /**
   * Execute command in sandbox or fallback
   */
  async run(command: string, timeout: number = 60): Promise<CommandResult> {
    if (this.isSandboxed && this.sandbox) {
      // Docker sandboxed execution
      try {
        const result = await this.sandbox.exec(command);
        return {
          stdout: result.stdout,
          stderr: result.stderr,
          returncode: result.exit_code,
          sandboxed: true,
        };
      } catch (error: any) {
        return {
          stdout: '',
          stderr: `Sandbox error: ${error.message}`,
          returncode: -1,
          sandboxed: true,
        };
      }
    } else if (this.fallback) {
      // Strict safety fallback
      const allowed = this.fallback.check(command);
      
      if (!allowed.allowed) {
        // Try to get explicit confirmation
        if (this.fallback.promptConfirmation(command)) {
          // Proceed with execution
        } else {
          return {
            stdout: '',
            stderr: `STRICT SAFETY BLOCKED: ${allowed.reason}`,
            returncode: -1,
            sandboxed: false,
          };
        }
      }

      // Execute with basic subprocess
      try {
        const { stdout, stderr } = await execAsync(command, {
          cwd: this.projectRoot,
          timeout: timeout * 1000,
        });
        return {
          stdout: stdout || '',
          stderr: stderr || '',
          returncode: 0,
          sandboxed: false,
        };
      } catch (error: any) {
        return {
          stdout: '',
          stderr: error.stderr || error.message,
          returncode: error.code || -1,
          sandboxed: false,
        };
      }
    }

    return {
      stdout: '',
      stderr: 'No execution method available',
      returncode: -1,
      sandboxed: false,
    };
  }

  /**
   * Cleanup sandbox resources
   */
  async stop(): Promise<void> {
    if (this.sandbox) {
      await this.sandbox.stop();
    }
  }
}

/**
 * RAG-based semantic search over codebase memory
 */
export class SearchMemoryTool {
  private memory: any; // RAGMemory type would be imported

  constructor(memory: any) {
    this.memory = memory;
  }

  searchCodebase(query: string, k: number = 5): string {
    if (!this.memory.available) {
      return '[Memory unavailable - RAG system offline]';
    }
    return this.memory.search(query, k);
  }

  getStats(): Record<string, any> {
    return this.memory.getStats();
  }
}

/**
 * Sandboxed filesystem operations
 */
export class FileSystemTool {
  root: string;

  constructor(root: string) {
    this.root = path.resolve(root);
    if (!fs.existsSync(this.root)) {
      fs.mkdirSync(this.root, { recursive: true });
    }
  }

  /**
   * Validate path is within root (prevents path traversal)
   */
  private validate(targetPath: string): string {
    const target = path.resolve(this.root, targetPath);
    const realRoot = fs.realpathSync(this.root);
    // Resolve symlinks to prevent traversal via symlinks
    let realTarget: string;
    try {
      realTarget = fs.realpathSync(target);
    } catch {
      // File may not exist yet (for write), check parent
      const parent = path.dirname(target);
      try {
        const realParent = fs.realpathSync(parent);
        if (!realParent.startsWith(realRoot)) {
          throw new Error(`Access denied: ${targetPath} outside project root`);
        }
      } catch {
        // Parent doesn't exist either, just check resolved path
        if (!target.startsWith(realRoot)) {
          throw new Error(`Access denied: ${targetPath} outside project root`);
        }
      }
      return target;
    }
    if (!realTarget.startsWith(realRoot)) {
      throw new Error(`Access denied: ${targetPath} outside project root`);
    }
    return realTarget;
  }

  private static readonly MAX_FILE_SIZE = 10 * 1024 * 1024; // 10MB
  private static readonly MAX_DIR_ENTRIES = 5000;

  /**
   * Read file contents
   */
  read(filePath: string, limit?: number): FileResult {
    try {
      const target = this.validate(filePath);
      if (!fs.existsSync(target)) {
        return { success: false, error: `Not found: ${filePath}` };
      }
      const stat = fs.statSync(target);
      if (!stat.isFile()) {
        return { success: false, error: `Not a file: ${filePath}` };
      }
      if (stat.size > FileSystemTool.MAX_FILE_SIZE) {
        return { success: false, error: `File too large (${Math.round(stat.size / 1024 / 1024)}MB > 10MB limit)` };
      }

      let content = fs.readFileSync(target, 'utf-8');
      if (limit && content.length > limit) {
        content = content.slice(0, limit) + `\n... (${content.length - limit} more)`;
      }
      return { success: true, content };
    } catch (error: any) {
      if (error.message.includes('Access denied')) {
        return { success: false, error: error.message };
      }
      return { success: false, error: `Read error: ${error.message}` };
    }
  }

  /**
   * Write file contents
   */
  write(filePath: string, content: string): FileResult {
    try {
      const target = this.validate(filePath);
      const dir = path.dirname(target);
      if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir, { recursive: true });
      }
      fs.writeFileSync(target, content, 'utf-8');
      return { success: true };
    } catch (error: any) {
      if (error.message.includes('Access denied')) {
        return { success: false, error: error.message };
      }
      return { success: false, error: `Write error: ${error.message}` };
    }
  }

  /**
   * List directory contents
   */
  listDir(dirPath: string = '.'): FileResult {
    try {
      const target = this.validate(dirPath);
      if (!fs.existsSync(target)) {
        return { success: false, error: `Not found: ${dirPath}` };
      }
      if (!fs.statSync(target).isDirectory()) {
        return { success: false, error: `Not a directory: ${dirPath}` };
      }

      const entries = fs.readdirSync(target);
      if (entries.length > FileSystemTool.MAX_DIR_ENTRIES) {
        return { success: false, error: `Directory has ${entries.length} entries (limit: ${FileSystemTool.MAX_DIR_ENTRIES})` };
      }
      const formatted = entries.map(e => {
        const fullPath = path.join(target, e);
        const isDir = fs.statSync(fullPath).isDirectory();
        return `${isDir ? '[dir]' : '[file]'} ${e}`;
      });

      return { success: true, content: formatted.sort().join('\n') };
    } catch (error: any) {
      if (error.message.includes('Access denied')) {
        return { success: false, error: error.message };
      }
      return { success: false, error: `List error: ${error.message}` };
    }
  }
}
